src/datasources/db/v2/database-migrator.service.ts

```diff
@@ -0,0 +1,160 @@
+import { PostgresDatabaseService } from '@/datasources/db/v2/postgres-database.service';
+import {
+  LoggingService,
+  type ILoggingService,
+} from '@/logging/logging.interface';
+import { Inject, Injectable } from '@nestjs/common';
+import { ConfigService } from '@nestjs/config';
+import type { DataSource } from 'typeorm';
+
+interface LockSchema {
+  id: number;
+  name: number;
+}
+
+enum MigrationStatus {
+  RUNNING = 1,
+}
+
+@Injectable()
+export class DatabaseMigrator {
+  public constructor(
+    @Inject(LoggingService) private readonly loggingService: ILoggingService,
+
+    @Inject(PostgresDatabaseService)
+    private readonly databaseService: PostgresDatabaseService,
+
+    @Inject(ConfigService)
+    private readonly configService: ConfigService,
+  ) {
+    //
+  }
+
+  /**
+   * Performs the migration operation.
+   *    Since pg_advisory_lock locks are session-based, they seem to be quite unreliable for our use case.
+   *    If the session is terminated, the lock is released, which could potentially cause issues in the database.
+   *    For that reason we are implementing a lock mechanism
+   *
+   * @returns {Promise<void>} A promise that resolves when the migration is complete.
+   *
+   * @throws {Error} Throws an error if unsuccessful after a specified number of retries."
+   */
+  public async migrate(): Promise<void> {
+    this.loggingService.info('Migrations: Running...');
+
+    const connection =
+      await this.databaseService.initializeDatabaseConnection();
+    await this.createLockTableIfNotExists(connection);
+
+    let numberOfIterations = 0;
+    const numberOfRetries = await this.configService.getOrThrow(
+      'db.migrator.numberOfRetries',
+    );
+    while (numberOfRetries >= numberOfIterations) {
+      ++numberOfIterations;
+      if (!(await this.lockExists(connection))) {
+        await this.insertLock(connection);
+        await this.runMigration(connection);
+        await this.truncateLocks(connection);
+
+        this.loggingService.info('Migrations: Finished.');
+        break;
+      } else {
+        if (numberOfIterations === numberOfRetries) {
+          throw new Error(
+            'Migrations: Migrations are still running in another instance!',
+          );
+        }
+        this.loggingService.info('Migrations: Running in another instance...');
+        const retryAfter = this.configService.getOrThrow<number>(
+          'db.migrator.retryAfter',
+        );
+
+        await new Promise((resolve) => setTimeout(resolve, retryAfter));
```

Comment:
Do we also need to cleanup this timeout if the app closes? It might cause open handles in our tests?

Reply:
Resolving after discussion: as we are breaking/throwing, this shouldn't be an issue. The tests have reduced `retryAfter`/`numberOfRetries` values.

Reply:
I think since we are breaking out of the loop either using `break` or `throw` that shouldn't be an issue. For tests I reduced the retryAfter and numberOfRetries .

Reply:
Naive question maybe: why do we need a retry mechanism? If the lock is taken, couldn't the instance just log that the lock is taken and ignore the migrations?